internal/ioutil/buffercloser.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ioutil
+
+import (
+	"bytes"
+	"io"
+	"sync"
+)
+
+var _pool = sync.Pool{
+	New: func() interface{} {
+		return &BufferCloser{
+			b: &bytes.Buffer{},
+		}
+	},
+}
+
+// BufferCloser is a light wrapping around the bytes.Buffer that implements
+// a "Close" method to return the buffer to a sync.Pool of buffers.
+type BufferCloser struct {
```

Comment:
This is another situation where you have a public struct that can be created without using the constructor. If someone does `bufferCloser := &ioutil.BufferCloser{}`, `b` will be nil which means `ReadFrom, WriteTo, Read` will panic, and even if `bytes.Buffer` was not a pointer, the value would not be initialized from the pool anyways, making `Close` not work as intended.

You can fix this in one of two ways:

- For every call to a function on `*BufferCloser`, check if `b` is properly initialized. This is expensive.
- Make `BufferCloser` a private struct, and instead return an interface that implements `io.ReaderFrom, io.WriterTo, io.Closer`. This is what I would recommend, I think the benefits outweigh the compatibility costs of using a struct.

All code in the Golang stdlib that exposes public structs ensures that there will not be issues regardless of if an user creates the struct by declaring it, or with the corresponding constructor function.

Reply:
Exposing public structs, even if the zero-value is not valid, is a very common
pattern used throughout YARPC, other libraries, and the standard library. The
expectation is that if a constructor is provided, unless otherwise documented,
you must use the constructor.

In the standard library at least, the documentation explicitly states that the
zero-value is valid if that's the case rather than letting you assume it.

Some structs in the standard library go out of their way to ensure that they
don't explode in the zero-value case (regexp.Regexp is definitely not one of
them) but I don't think it's worth the effort or performance cost here.

Reply:
I can't think of an example in the stdlib that does not allow you to use a public struct you just declare, they go out of their way to make sure they work no matter how you declare them, as they should - you're knowingly publicly exposing something that could panic otherwise. If you can think of an example that documents the zero-value is invalid, I'd love to see it, as I've seen comments in the stdlib that say things like "we do this because if someone declares the struct instead of using the constructor it doesn't work".

tl;dr I don't know of an example in stdlib that will panic in this case, and regardless it's a bad pattern to expose something that can error. Hence why it's usually nicer to expose an interface here, or to handle the error cases.

Reply:
> If you can think of an example that documents the zero-value is invalid [..]

Actually, the opposite. Structs explicitly state that their zero value is valid. So you could say that the default is invalid unless stated otherwise.

Simple example of a struct that's invalid for zero: https://golang.org/pkg/regexp/#Regexp. Panic demonstration: https://play.golang.org/p/NQg1akM6iS.

Reply:
`Regexp` is a special case as how they are constructed is unusual, and I wouldn't agree with that anyways. `bytes.Buffer` is a much more normal case here, as is `http.Server`, etc. All of these work to make sure the zero-value works, as they should.

Reply:
More examples:

- json.Decoder: https://play.golang.org/p/YagPipX545
- log.Logger: https://play.golang.org/p/EOdlj5nWeL
- reflect.Value is all panics but that's probably a special case

I agree that zero-value-just-works is a nice to have but when there is *some* required input, you can't have it. Again, [Accept interfaces return structs](http://idiomaticgo.com/post/best-practice/accept-interfaces-return-structs/) is pretty idiomatic to Go, including when the returned struct doesn't have a valid zero-value.

Reply:
Yea I get it, ok let's discuss at standup because my personal opinion is that it's really not good to expose something that can panic with incorrect usage (at the very least, `BufferCloser` should document this in the Godoc I'd say right?), I want us to be consistent as a team going forward.

Reply:
Actually just going to add this to weekly agenda for Thursday, let's discuss then, unblocking this for now.